cmd/promtool/analyze.go

```diff
@@ -0,0 +1,515 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/regexp"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+var (
+	errNotNativeHistogram = fmt.Errorf("not a native histogram")
+	errNotEnoughData      = fmt.Errorf("not enough data")
+
+	sumSuffix = regexp.MustCompile(`_sum\b`)
+)
+
+type QueryAnalyzeConfig struct {
+	metricType string
+	duration   time.Duration
+	time       string
+	matchers   []string
+}
+
+// run retrieves metrics that look like conventional histograms (i.e. have _bucket
+// suffixes) or native histograms, depending on metricType flag.
+func (c *QueryAnalyzeConfig) run(url *url.URL, roundtripper http.RoundTripper) error {
+	if c.metricType != "classichistograms" && c.metricType != "nativehistograms" {
+		return fmt.Errorf("analyze type is %s, must be 'classichistograms' or 'nativehistograms'", c.metricType)
+	}
+
+	ctx := context.Background()
+
+	api, err := newAPI(url, roundtripper, nil)
+	if err != nil {
+		return err
+	}
+
+	var endTime time.Time
+	if c.time != "" {
+		endTime, err = parseTime(c.time)
+		if err != nil {
+			return fmt.Errorf("error parsing time '%s': %w", c.time, err)
+		}
+	} else {
+		endTime = time.Now()
+	}
+
+	if c.metricType == "nativehistograms" {
+		return c.getStatsFromMetrics(ctx, api, endTime, os.Stdout, identity, true, c.matchers)
+	}
+	for _, matcher := range c.matchers {
+		if !sumSuffix.MatchString(matcher) {
+			return fmt.Errorf("every classic histogram matcher must have a '_sum' suffix in the metric name, but the matcher '%s' does not", matcher)
+		}
+	}
+	return c.getStatsFromMetrics(ctx, api, endTime, os.Stdout, replaceSumWithBucket, false, c.matchers)
```

Comment:
We search for _sum as that will only have 1 series, while _bucket will have many, and we use it as kind of the index for subsequent searches to break it down into smaller queries. If we decide to go with your one-query approach, we will remove this part, but if we keep this approach, we can easily let them search _bucket instead if that is more consistent for the user (but internally we will search _sum for the index part).

Reply:
I would suggest to first get an MVP going that is based on a single query. Then we can play with it and see what kind of analysis is useful and what the potential problems are with too many time series to process.

I see some problems with the approach as implemented in the current state of the PR:
- It's inconsistent between native and classic histograms. (You could do a 2-step approach with native histograms, too, maybe using the [series endpoint](https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers), which might, by the way, also be more appropriate for the current classic histogram approach.)
- It is more complex and more expensive in the common case where we can handle all the buckets at once. (A typical scenario would be hundreds of histograms with only a handful buckets each. This would require hundreds of round trips rather than one. Things might take unbearably long then.)
- It relies on the existence of the `_sum` series, which is not guaranteed. (In fact, the `_count` series would be more likely to exist, but even that cannot be guaranteed.)
- It requires the user to provide the matcher in a specific way. For example, the user might just provide `{foo="bar"}` because they know that all series matched by this label are histograms (and even if they aren't, we could still filter for `_bucket` series). This would not work with the current code, and the reason would be hard to understand for the user.

If we really need an approach that distributes the work load over multiple queries, then I would expect it to be a bit more involved to be feasible, like not fetching every single histogram separately (of which there could be thousands), but fetch them in larger batches.

The reason for my suggestion to go with the single query approach first is that it solves the problem for the case where we do not have to process too many histograms, and it even gives the user the opportunity to divide the work load manually by running the command multiple times with hand-crafted matchers. All of this while avoiding all the complications above. Therefore, we would have code that we can merge and which then could be used and on which we can iterate without breaking what we have merged already.

Reply:
> You could do a 2-step approach with native histograms, too, maybe using the [series endpoint](https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers), which might, by the way, also be more appropriate for the current classic histogram approach.

But you'd prefer single query over using this series endpoint first?

Reply:
Yes. I propose to first implement the single query approach consistently (because we know it will work for common cases). Then we get this merged, and then we can iterate on more involved approaches where needed.